### ISSUE TITLE
can't make this work with VSCode 1.28.2

### ISSUE THREAD

**Issue by davydden** [paulo-fernando-silva/vscOctaveDebugger#9]
I was looking for Octave debugger and came across this pluging. However, I can't make it work on macOS with VSCode 1.28.2. Issues so far:

- [ ] no red bullet points to set for each line in `.m` file
- [ ] I setup a launch target
```
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "type": "OctaveDebugger",
      "request": "launch",
      "name": "Octave",
      "program": "${workspaceFolder}/test",
      // "program": "${workspaceFolder}/${command:AskForProgramName}",
      "stopOnEntry": true,
      "octave": "/full/path/to/bin/octave-cli",
      "sourceFolder": "${workspaceFolder}"
    }
  ]
}
```
but trying to run debugger does not give anything. I don't see anything in logs either.

**Comment by PeterEnis1234** [comment]
Same problem on Windows

**Comment by paulo-fernando-silva** [comment]
Hey Denis, can you try setting
"octave": "/full/path/to/bin/octave-cli",
to for example
"octave": "octave-cli"
and let me know if that works?
I still need to add some feedback console logging for errors.

**Comment by davydden** [comment]
@paulo-fernando-silva 

>  "octave-cli" on your path.

hm, why is that? I prefer NOT to have it in my path and it's not how I use it/install it. I would expect that the full path to the binary should work.

Anyway, I added it to the path and it does NOT make a difference, unfortunately...

> did you install it from within VSC plugins tab or from source?

you mean octave? no I build it via different tools and it works from terminal.
if you mean the debugger, yes, I installed it via the plugins tab.

> That picks up the function test located under "${workspaceFolder}" and subdirectories, or under whatever directories you have setup in your octave default search path.

so you are saying that `program` is a relative path, also without the `.m`? Perhaps it would be a good idea to document launch configuration options for this debugger. 

p.s. full path or just executable name should be a separate issue to the issue that one can't set debug points.

p.p.s. are you aware of https://github.com/gitletH/vscode-octave-debug ? Perhaps you guys can join the forces

**Comment by paulo-fernando-silva** [comment]
Hey Denis,

yes, the full path should work. I tried it just now and it does work for me.
I'm imagining it's something else. I added some logging which was missing and I didn't even noticed that the program output wasn't shown on the console.
Now both program output and debugger communication can be seen.
To activate the debugger communication you can set trace to true. But you'll probably won't need it unless you're debugging the extension.


Actually, I imagined that you were using
	"octave": "/full/path/to/bin/octave-cli",
as the parameter value, but I believe you don't have octave installed under "/full/path/to/bin/octave-cli".
But anyway, the current version (v0.2.1) now produces the following default configuration:
        {
            "type": "OctaveDebugger",
            "request": "launch",
            "name": "debug main",
            "program": "main",
            "stopOnEntry": true,
            "octave": "octave-cli",
            "sourceFolder": "${workspaceFolder}",
            "prefetchCount": 1,
            "allowArbitraryExpressionEvaluation": false,
            "trace": false
        }

I also tried with
            "octave": "/opt/local/bin/octave-cli",
which is where I have my octave installed, and it works fine. But let me know if it still doesn't work for you. Now if it fails to open a pipe to the octave process it'll output and error saying so.



Well, as for program it can be anything, e.g.
            "program": "/Users/paulo/dev/matlab/lms/main.m",
will call and run the following:
			octave> addpath("/Users/paulo/dev/matlab/lms/");
			octave> main
As for
            "sourceFolder": "${workspaceFolder}",
it's basically redundant, but it helps if you want to run a function such as 
        "program": "/Users/paulo/dev/matlab/lms/main.m",
but you want the source folder to be
            "sourceFolder": "/Users/paulo/dev/matlab",

Since including 
	"/Users/paulo/dev/matlab"
also includes
	"/Users/paulo/dev/matlab/lms/"
You can basically just set "sourceFolder" and call any function under that directory by name.


Yes, the exe name is indeed a separate issue from setting breakpoints. I tried to reproduce your issue but I wasn't able to. I'm also running vsc 1.28.2 and it sets breakpoints even if I set the the octave binary to something random. I need more info. Maybe the logging will help us find the issue.

I do know about the other octave debugger. I wish he had done it before I stated mine. I've been implementing the debugger interface when all I wanted was to program in matlab/octave. Anyway, even though everyone uses python these days, matlab is still great for prototyping, so having a debugger in my IDE of choice is really handy. Having two is perhaps better. But be warned that as soon as I have this working as I want, which is about done, I probably won't implement any features or provide time-consuming support. I just don't have the time. Obviously anyone that thinks the code is useful is free to clone and run with it. The whole point is to give back to the community.

Just try the new version and let me know if it works for you.

**Comment by davydden** [comment]
> Actually, I imagined that you were using
"octave": "/full/path/to/bin/octave-cli",
as the parameter value, but I believe you don't have octave installed under "/full/path/to/bin/octave-cli".

nah, it is installed in my path, no mistake here.

ok, here is the root for the debugger part of the issue:

```
error: remove_all_breakpoint_in_file: unable to find function foo.cc
```

that's because I have **disabled** breakpoint in some `foo.cc` file I used a while ago for C++. The file **does** exist, so it should be valid to keep breakpoints in other workspaces.

There must be a way to only take breakpoints from current workspace or preferably only from `.m` files within the workspace (sorry, i know that it's possible as that other debugger project works in the same setting, but my impression was that it has fewer features, so I would like to figure out what is the reason for these issues here...).

If I remove that C++ breakpoint altogether, then debugger actually starts. But setting breakpoints still is not functional.

p.s. After trying to run a debugger, `octave-cli-4.4.1` process is alive and takes `99.9%` of CPU and 3 threads. I know nothing about debugger adapters in VSCode, but perhaps you could kill it somehow...

> Maybe the logging will help us find the issue.

how can I try to log this GUI part of the issue?

**Comment by paulo-fernando-silva** [comment]
I'm not sure if I understand the issue. You're saying that you have a foo.cc in your matlab code workspace? I'm imagining you have two more than one project (directory) opened in the same workspace, and if I understand correctly the debugger is picking up the breakpoints from an unrelated project and trying to set them in the current (matlab) project. Is that it?

Not sure why the breakpoints aren't working. Could you try a simple 1 directory project? Just for testing.

I've created a new bug for the octave-cli process that doesn't sometimes doesn't exit. I know about the issue, but still didn't have time to fix it. Usually it happens when you terminate the debug session in the middle of a breakpoint. Should be simple to fix.

Which GUI part do you want to debug? To debug the extension output just set  "trace": true in the launch.json

**Comment by davydden** [comment]
> I understand correctly the debugger is picking up the breakpoints from an unrelated project and trying to set them in the current (matlab) project. Is that it?

yes

> Not sure why the breakpoints aren't working. Could you try a simple 1 directory project? Just for testing.

I can set `octave` breakpoints in another debugger project. So it's certainly not related to workspace.

> Which GUI part do you want to debug?

breakpoints.

> To debug the extension output just set "trace": true in the launch.json

that does not have anything to do with debugging, right? It's about not being able to set breakpoints **before** running debugger.

**Comment by paulo-fernando-silva** [comment]
Ok, got it. Both are really strange.
Can you make me a simple project that reproduces the issues that you're having?
Not sure how to debug it any other way.

You can try to select the correct debugger profile from the dropdown box under the debugger cmd+alt+D.

**Comment by davydden** [comment]
> You can try to select the correct debugger profile from the dropdown box under the debugger cmd+alt+D.

it's the correct profile. It's just I have other folders in the workspace, some with C++ projects and there I have breakpoints for `C/C++` plugging from Microsoft.

> Can you make me a simple project that reproduces the issues that you're having?

I'll try to do that...

**Comment by davydden** [comment]
@paulo-fernando-silva 

> Can you make me a simple project that reproduces the issues that you're having?

ok, here's a reproducer for you.

- 2 files in two folders.

```
/c++
/c++/hallo.c
/octave
/octave/hallo.m
```

- install C/C++ plugin from Microsoft
- set breakpoint in `.c` file (you should see it in the list bottom left of breakpoint, recall that I can NOT set breakpoints in `.m`)
- add debug configuration for your Octave plugin (I kept all default) and try to run it (you do NOT need to setup any debug config for `octave`)
- see the attached screenshot for error

<img width="1458" alt="screenshot 2018-11-04 at 21 22 25" src="https://user-images.githubusercontent.com/8023934/47969476-89f48180-e078-11e8-8c50-38c5760c0072.png">

As I said, my best guess is that your plugin simply try to use all breakpoints set everywhere within the workspace.

**Comment by paulo-fernando-silva** [comment]
I tried a quick fix (only accepting .m), but it doesn't seem to jive with vscode.
I have to get back to this at a later time. In the meanwhile I asked a question here:
https://stackoverflow.com/questions/53158385/how-to-properly-handle-breakpoints-in-multiroot-workspaces
Also, since this issue is unrelated to the original issue, I believe this issue is fixed, and so I'll close it.
I created a new issue here 
https://github.com/paulo-fernando-silva/vscOctaveDebugger/issues/15
